### PR TITLE
Add sequence view

### PR DESCRIPTION
**Pull request by jakobnissen**
This adds a sequence view `SeqView` as discussed in #102 .

Because a `LongSequence` and `SeqView` share the same underlying encoding, many of the old `LongSequence` methods has been changed to now take a `SeqOrView` (which is just `Union{LongSequence, SeqView}`)

You can construct a view with
* `SeqView(s::LongSequence, ::UnitRange)`
* `@view my_seq[1:5]`
* `SeqView(::Vector{UInt64}, ::UnitRange)` (this one doesn't boundscheck)

Mutating methods that change the length of the view are intentionally not implemented, such as
* `filter!`
* `map!`
* `resize!`
* `ungap!`

### CODE REVIEW

**Comment by BenJWard** [comment]
Is editing a `LongSequence`, making it smaller say as a result of `ungap!` or `resize!` going to present a problem for a previously created view of say `@view seq[90:100]` if the sequence is resized to only 90bp?

File: src/longsequences/transformations.jl [hunk 1]
```diff
@@ -11,50 +11,40 @@ array unless the new size does not fit. If `force`, always resize underlying dat
...
-            if index(next) != j
-                seq.data[j] = datum
-                datum = 0
-                j = index(next)
+function Base.filter!(f, seq::LongSequence)
+    writeindex = bitindex(seq, 0)
+    chunk = zero(UInt64)
+    bps = bits_per_symbol(seq)
+    for i in eachindex(seq)
+        encoded_symbol = extract_encoded_element(bitindex(seq, i), encoded_data(seq))
```
Comment (BenJWard) [inline src/longsequences/transformations.jl:28]: Doesn't inbounds_getindex do these two lines already?

Comment (jakobnissen) [reply src/longsequences/transformations.jl:28]: Yep, but I'm using the encoded symbol a few lines down, and don't want to redo the calculation,

**Comment by jakobnissen** [comment]
> Is editing a `LongSequence`, making it smaller say as a result of `ungap!` or `resize!` going to present a problem for a previously created view of say `@view seq[90:100]` if the sequence is resized to only 90bp?

Yes, the view is completely unsafe. In theory it could even segfault if the underlying array is truncated. Perhaps I could add another boundscheck when accessing a view that the underlying buffer must be long enough? That would slow down boundschecked access just a little.

**Comment by BenJWard** [comment]
Ok, I'm happy with whatever behaviour closest resembles what julians already expect from a view, since that will be most consistent. We'll just make sure to document any pitfalls properly.

**Comment by CiaranOMara** [comment]
I was proposing "Sub" as a prefix convention.

**Comment by jakobnissen** [comment]
I think this has been hanging around long enough - I'm merging this now, and will begin using `master` branch as my daily driver. Hopefully I can hammer out any remaining bugs that way.

**Pull request merged**